Automatically add font license info for Google fonts

mikachan opened the pull request:
This adds a function that automatically handles adding the font license information for each Google font to the theme's readme.txt file.

<img width="681" alt="image" src="https://user-images.githubusercontent.com/1645628/229370361-92ff65b7-0649-407e-b92f-b66d5e56d403.png">

Part of https://github.com/WordPress/create-block-theme/issues/195.

mikachan committed:
Add font license function for Google fonts

mikachan committed:
Install php-font-lib

mikachan committed:
Use php-font-lib to check for license info

mikachan committed:
Use custom format for OFL

mikachan committed:
Rename function to manage_font_license

Handle remove as well as add functionality

mikachan committed:
Squashed merge commit with trunk

commit 0374c6974ee25671557d6fcebbd4cf9db436a409
Merge: 83de623 072da09
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Tue Apr 4 16:15:50 2023 +0100

    Merge pull request #297 from WordPress/add/font-family-labels

    Add labels around Google font family checkbox controls

commit 83de62353684a92d1da7163a2aff956615603ee9
Merge: 58491a7 b814f29
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Tue Apr 4 16:13:48 2023 +0100

    Merge pull request #296 from WordPress/try/add-current-wp-version

    Add current WordPress version to style.css and readme.txt

commit 58491a7618e57fd8c266f8912cbfdcfc43699f46
Author: github-actions <github-actions@github.com>
Date:   Mon Apr 3 20:54:08 2023 +0000

    Version bump & changelog update

commit 3a58a30cec13f9e3339d65a73e3cbf9b5116b215
Merge: bb36a3a fd779d3
Author: Jeff Ong <jonger4@gmail.com>
Date:   Mon Apr 3 11:27:38 2023 -0400

    Merge pull request #292 from WordPress/try/site-editor-overwrite

    Add Export (Clone) to site editor

commit fd779d30b1b17b1101ef4d1b3f488dedb2c121f8
Author: Jeff Ong <jonger4@gmail.com>
Date:   Mon Apr 3 11:26:11 2023 -0400

    Carry forward previous theme tags.

commit 072da09df4ffebebb53c5cd817908545b6e097b2
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Sat Apr 1 21:39:54 2023 +0100

    Add checkbox labels to font family table header

commit 54318c68743edbc2216cf1e10b4c32ef9c20b011
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Sat Apr 1 21:39:22 2023 +0100

    Add checkbox labels to font family variants

commit b814f291041ac908426501f2eb3acfa06549b006
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Sat Apr 1 20:54:04 2023 +0100

    Add current WP version to readme.txt

commit f4070b2522e63d4a33b88aa87957c1a5b724e564
Author: Sarah Norris <sarah@sekai.co.uk>
Date:   Sat Apr 1 20:53:50 2023 +0100

    Add current WP version to style.css

commit bb36a3adbd44791a2e7cc67ec73997d45605b28b
Merge: b5a0a0a 8c4098a
Author: Matias Benedetto <matias.benedetto@gmail.com>
Date:   Fri Mar 31 17:35:57 2023 -0300

    Merge pull request #293 from WordPress/update/google-fonts-json-_4537839996

    Update Google Fonts JSON data from API

commit cd5f862db849f65897ffb485809cf2d695c04634
Author: Matias Benedetto <matias.benedetto@gmail.com>
Date:   Fri Mar 31 17:22:36 2023 -0300

    setting the filename of the exported zip file

commit b5a0a0aeca3addcd3608b91c669752f52220d0e0
Merge: e3f9193 260d115
Author: Matias Benedetto <matias.benedetto@gmail.com>
Date:   Fri Mar 31 13:05:52 2023 -0300

    Merge pull request #291 from WordPress/fix/280

    Fixing image downloading not working in some cases

commit 1b19d29e297e919172514266d7afbbe2da092db4
Author: Jeff Ong <jonger4@gmail.com>
Date:   Thu Mar 30 12:24:35 2023 -0400

    Make linter happy.

commit b86a1f12f0c89ef716ac087e8a7593ca34e4790c
Author: Jeff Ong <jonger4@gmail.com>
Date:   Thu Mar 30 12:23:42 2023 -0400

    Rename file.

commit 6efb440d1cafcff3ffd1a8e4548227ab29eb0a80
Author: Jeff Ong <jonger4@gmail.com>
Date:   Wed Mar 29 17:27:42 2023 -0400

    Add validation to form button, improve spacing.

commit b5c6a17d5321c2618b363ce38fef49f8fcb7cf6a
Author: Jeff Ong <jonger4@gmail.com>
Date:   Wed Mar 29 17:11:01 2023 -0400

    Async/await, update icon, button.

commit 8bc38efd9ff3d44844fe4001f8e4d812a4f194e5
Author: Jeff Ong <jonger4@gmail.com>
Date:   Wed Mar 29 16:40:00 2023 -0400

    Use tt3 as placeholder

    Co-authored-by: Ben Dwyer <ben@scruffian.com>

commit 16f3c5cc0f59955498eaccf8d367389ebab22620
Author: Matias Benedetto <matias.benedetto@gmail.com>
Date:   Wed Mar 29 15:50:17 2023 -0300

    Implementing WordPress apiFetch replacing the browser's fetch function

commit 4870efb7f3e640681ccb1ada9190cc6308c0c176
Author: Jeff Ong <jonger4@gmail.com>
Date:   Mon Mar 27 17:47:44 2023 -0400

    Lint php.

commit ebdf773b1556955386aeefb4dbc6ddffe2fe9070
Author: Jeff Ong <jonger4@gmail.com>
Date:   Mon Mar 27 17:47:22 2023 -0400

    Site editor rest API export.

commit a7b2de44f5cc9c2450ec1899ddb93e5ad6b9709e
Author: Jeff Ong <jonger4@gmail.com>
Date:   Mon Mar 27 13:49:35 2023 -0400

    Add panel to sidebar in site editor.

commit 8c4098a514ace8ce46cc834dfdedca86f12b899d
Author: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Date:   Tue Mar 28 00:15:24 2023 +0000

    Updating file

commit 260d1159126ecd1cde12745139e522635285a197
Author: Matias Benedetto <matias.benedetto@gmail.com>
Date:   Mon Mar 27 15:48:44 2023 -0300

    fixing image downloading not working in some cases

mikachan committed:
Try to fix fallback-fonts-list file

mikachan committed:
Revert "Try to fix fallback-fonts-list file"

This reverts commit 8d05847d9311f467c481e18ecc251640be12db13.

mikachan committed:
Revert "Squashed merge commit with trunk"

This reverts commit a4397ded1ebd61850c369f595ac44417d730a55a.

mikachan committed:
Replace php-font-lib with a later release

mikachan committed:
Rename font name variable

mikachan committed:
Use stripos rather than strpos

mikachan committed:
Refactor remove functionality

matiasbenedetto commented on admin/class-manage-fonts.php:
```diff
@@ -296,7 +305,82 @@ function ( $font_family ) use ( $font_slug ) {
+				$font->parse();
+				$license_info = $font->getNameTableString( 13 ) ? $font->getNameTableString( 13 ) : '';
+
+				if ( stripos( $font->getNameTableString( 13 ), 'SIL Open Font License' ) ) {
```
Could get all the information from the file metadata instead of searching for specific things?

mikachan replied:
I'm not sure what you mean here, could you give me an example? This is to check if the SIL license is being used, and then replace the information from the file with a custom message so it's more consistent with the other types of license.

matiasbenedetto replied:
I mean that we can extract the entire license information from the metadata tables of the font file asset. 
Example:
![image](https://user-images.githubusercontent.com/1310626/232865567-5ead5fdf-716d-47ea-8e11-983c8a912311.png)

This online font metadata inspector could be helpful in looking for the name of the metadata property: https://opentype.js.org/font-inspector.html

mikachan replied:
Ah yes, that's what I'm using the FontLib library for here, but the information for the SIL license is a lot longer and looks different from the usual descriptions we use for the fonts in WP themes. So I decided to add a check for the SIL license and then replace it with information in our usual format.

matiasbenedetto reviewed (commented):
Thanks for working on this! Very useful enhancement! :heart_eyes: 

My main concern is this: We already have a client-side library to read the font metadata. Do you think it is worth bundling another entire library for the backend? 

I don't have an answer to that question, but maybe this could help to decide:
Would the use of just the client-side library to extract the metadata increase or decrease code complexity?

matiasbenedetto commented:
When I add a font, 3 deprecation notices are displayed:

```
Deprecated: file_get_contents(): Passing null to parameter #4 ($offset) of type int is deprecated in create-block-theme/includes/FontLib/Font.php on line 31

Deprecated: Function strftime() is deprecated in /create-block-theme/includes/FontLib/BinaryStream.php on line 285

Deprecated: Function strftime() is deprecated in create-block-theme/includes/FontLib/BinaryStream.php on line 285

```

matiasbenedetto commented:
Not all the fonts available on Google Fonts are licensed under SIL license, there are a few licensed under Apache or Ubuntu licenses. If I add for example the font called Arimo, it adds a SIL license notice when Arimo uses Apache licensing. https://fonts.google.com/attribution

Example:
```
Arimo Font
Licensed under the SIL Open Font License, Version 1.1 (http://scripts.sil.org/OFL)
Source: https://www.google.com/fonts/specimen/Arimo
```

matiasbenedetto commented:
Should we add the copyright notice and the author's name to be compliant with the attribution part of the SIL license? I guess that it won't hurt and we can we that info [from the file metadata](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6name.html). In this way, we could make this work for local font files too.

Apparently, we need to add the copyright notice too: https://themes.trac.wordpress.org/ticket/136638#comment:2

matiasbenedetto commented on admin/class-manage-fonts.php:
```diff
@@ -296,7 +305,82 @@ function ( $font_family ) use ( $font_slug ) {
+				require_once( __DIR__ . '/../includes/FontLib/Autoloader.php' );
+
+				// Build font source URL
+				$font_source = 'https://www.google.com/fonts/specimen/' . $google_formatted_font_name;
```
Could we get rid of this and use the info from the font file asset metadata? In this way we could use this function for local font files too.

mikachan replied:
Yes good idea, I was hoping to extend this to local fonts once I got this version working ✨

matiasbenedetto commented on composer.json:
```diff
@@ -32,12 +32,16 @@
-		"test:watch": "phpunit-watcher watch < /dev/tty"
+		"test:watch": "phpunit-watcher watch < /dev/tty",
+		"post-autoload-dump": [
+			"cp -r vendor/phenx/php-font-lib/src/FontLib includes"
```
I'm not sure what is this doing. Could you explain it, please :pray: ?

mikachan replied:
This is copying the FontLib files to the `includes` directory, as this seemed like a good way to include the library and expose the functionality. It's the first time I've needed to include an external library like this though, I'm not sure if this is the best way!

madhusudhand commented on admin/class-manage-fonts.php:
```diff
@@ -296,7 +305,82 @@ function ( $font_family ) use ( $font_slug ) {
+					$license_info = 'Licensed under the SIL Open Font License, Version 1.1';
+				}
+
+				$license_url = $font->getNameTableString( 14 ) ? "({$font->getNameTableString(14)})" : '';
```
Should we include author information to give them credits?

<img width="268" alt="image" src="https://user-images.githubusercontent.com/1935113/231711678-69583778-d00d-4921-ab6d-af16851c3b23.png">

mikachan replied:
Yes this is a good idea, I'll try adding it from the metadata!

madhusudhand reviewed (commented):
update `gitignore` to exclude `includes` dir?

mikachan commented:
> My main concern is this: We already have a client-side library to read the font metadata. Do you think it is worth bundling another entire library for the backend?

I'd rather not include a whole additional library that does the same thing, so I'd like to try and use the client-side library here instead. Hopefully it doesn't increase the code complexity too much.

mikachan commented:
> Not all the fonts available on Google Fonts are licensed under SIL license, there are a few licensed under Apache or Ubuntu licenses. If I add for example the font called Arimo, it adds a SIL license notice when Arimo uses Apache licensing.

This was the main reason I added the FontLib library 😅 Hmm, I was testing this using the Aclonica font, and the Apache license was coming through correctly from the metadata. Maybe this information isn't always correct with Google fonts... I'll do some more testing around this. Maybe it'll work better with the other library.

mikachan committed:
Use lib-font to get font credits

mikachan committed:
Use font-credits data passed from client-side

mikachan committed:
Remove php-font-lib

mikachan committed:
Add ending tag to credit info

mikachan committed:
Use array_key_exists consistently

mikachan commented:
Thank you @matiasbenedetto @madhusudhand for the reviews!

I've updated the PR with the following:

- Tried out using the client-side font library to read the font files and removed the PHP font library
- It's now passing the font credit info from the font metadata through the Google font form, along with all the other existing font data (name, weight etc)
- Included the copyright info
- It's now using the source URL from the font metadata
- Added an 'ending tag' to the font credit info in the readme.txt, as this makes it a lot easier to remove the credits when the font is removed from the theme

I think this covers the feedback so far. I prefer using a single font library!

matiasbenedetto reviewed (commented):
Adding some particular fonts, for example `Abyssinica SIL`, the entire SIL license text is printed in the `readme.txt` file. Is this expected?

![image](https://user-images.githubusercontent.com/1310626/234877722-22086a7b-59e5-4570-861c-45408d423402.png)

mikachan commented:
Thanks for testing again!

> Is this expected?

Ah, no! I've not seen that before. I'll look into how we can handle that.

mikachan committed:
Merge branch 'trunk' into add/handle-google-font-credits

mikachan committed:
Use get() function to get name records

mikachan committed:
Trim whitespace from all credit fields

mikachan committed:
Prevent license info from being over 200 chars

mikachan commented:
I think I've improved this with a bit of a refactor. 

- I've changed this so it's using the `get()` function to get the data from the font file, and this seems more reliable than before.
- I've limited the license info content to 200 characters, as this seemed reasonable. The full license can always be read at the license URL, which we're also including.

mikachan committed:
Refactor end of credits note

matiasbenedetto commented:
> I've limited the license info content to 200 characters

I haven't read the code yet, but can it be a good idea to limit it to one line instead of counting the characters?

mikachan committed:
Split license info at first new line

mikachan commented:
> but can it be a good idea to limit it to one line instead of counting the characters?

I considered this too, and couldn't decide which one was best! I've updated it so it will now limit the content to the first line, and still limit it to 200 characters as well in case there's one long single-line string.

matiasbenedetto reviewed (changes requested):
I found a little problem: let's say a family has several font faces; if you delete just one font face, but the rest of the family is there, the license is gone from `readme.txt` . Apart from that this PR seems almost ready to go!

Github video upload is not working for me today so I upload the screencast of featuring the issue here:
https://i.imgur.com/nNjO55W.mp4

mikachan committed:
Add new line above credits

mikachan committed:
Move delete functionality to prepare_font_families_for_database

mikachan commented:
Good spot, @matiasbenedetto! I've fixed this by moving the 'remove' functionality to the `prepare_font_families_for_database` function, so it can be run based on the font family logic rather than the font face. Thanks!

matiasbenedetto reviewed (approved):
It seems to be working great now. Thanks for this enhancement. Very useful.

As a follow up, I created an issue to extend this functionality to the local font file assets uploaded: https://github.com/WordPress/create-block-theme/issues/348

LGTM :rocket:
